src/__tests__/resolve-info.test.ts

```diff
@@ -0,0 +1,803 @@
+        Object {
+          "Bar1": Object {
+            "b1": true,
+            "id": true,
+            "title": true,
+          },
+          "Bar2": Object {
+            "b2": true,
+            "id": true,
+            "title": true,
+          },
+          "IBar": Object {
+            "id": true,
+            "title": true,
+          },
+        }
+      `);
+    });
+
+    test("aggregate multiple selections of the same field", async () => {
+      const result = await executeQuery(
+        schema,
+        parse(
+          `
+            query {
+              iBar {
+                id
+              }
+              iBar {
+                title
+              }
+              iBar {
+                ... on Bar1 {
+                  b1
+                }
+              }
+              iBar {
+                ... on Bar2 {
+                  b2
+                }
+              }
+            }
+          `
+        )
+      );
+
+      expect(result.errors).not.toBeDefined();
+
+      expect(inf.fieldExpansion).toMatchInlineSnapshot(`
+        Object {
+          "Bar1": Object {
+            "b1": true,
+            "id": true,
+            "title": true,
+          },
+          "Bar2": Object {
+            "b2": true,
+            "id": true,
+            "title": true,
+          },
+          "IBar": Object {
+            "id": true,
+            "title": true,
+          },
+        }
+      `);
+    });
+  });
+
+  describe("unions", () => {
+    let inf: any;
+    const schema = makeExecutableSchema({
+      typeDefs: `
+        type Query {
+          uBaz: Baz
+        }
+        union Baz = Foo | Bar
+        type Foo {
+          foo: String
+        }
+        type Bar {
+          bar: Int
+        }
+      `,
+      resolvers: {
+        Query: {
+          uBaz(_: any, _1: any, _2: any, info: any) {
+            inf = info;
+          }
+        },
+        Baz: {
+          __resolveType() {
+            return "Foo";
+          }
+        }
+      }
+    });
+
+    afterEach(() => {
+      inf = undefined;
+    });
+
+    test("union field nodes", async () => {
+      const result = await executeQuery(
+        schema,
+        parse(
+          `
+            query {
+              uBaz {
+                ... on Foo {
+                  foo
+                }
+                ... on Bar {
+                  bar
+                }
+              }
+            }
+          `
+        )
+      );
+
+      expect(result.errors).not.toBeDefined();
+      expect(inf.fieldExpansion).toMatchInlineSnapshot(`
+        Object {
+          "Bar": Object {
+            "bar": true,
+          },
+          "Foo": Object {
+            "foo": true,
+          },
+        }
+      `);
+    });
+
+    test("__typename", async () => {
+      const result = await executeQuery(
+        schema,
+        parse(`
+          query {
+            uBaz {
+              __typename
```

Comment:
Why are we excluding `__typename`?

Reply:
It was harder to handle this specific case for unions where it behaves like an interface. It's like in GraphQL the composite types are implicitly part of a super interface like the following - 

```graphql
interface __Typename {
  __typename: String;
}
```

and the fact that a union implements this `__Typename` just messes around with the notion of a selection field inside of a selectionSet. If we get new internals, it's manual effort to handle it.

And I didn't see any value in including it as a part of the field selection for the use case we have.   The use-case of `if __typename field is requested; then do things differently` isn't important and we do not see such things.